ibis/backends/datafusion/__init__.py

```diff
@@ -151,13 +121,86 @@ def register(
             Datafusion-specific keyword arguments
         """
         if isinstance(source, (str, Path)):
-            method, path, name = _register_file(
-                str(source), table_name=table_name, **kwargs
-            )
-            getattr(self._context, method)(name, str(path), **kwargs)
+            first = str(source)
         else:
             raise ValueError("`source` must be either a string or a pathlib.Path")
-        return self.table(name)
+
+        if first.startswith(("parquet://", "parq://")) or first.endswith(
+            ("parq", "parquet")
+        ):
+            return self.read_parquet(source, table_name=table_name, **kwargs)
+        elif first.startswith(("csv://", "txt://")) or first.endswith(
+            ("csv", "tsv", "txt")
+        ):
+            return self.read_csv(source, table_name=table_name, **kwargs)
+        else:
+            self._register_failure()
+            return None
+
+    def _register_failure(self):
+        import inspect
+
+        msg = ", ".join(
+            m[0] for m in inspect.getmembers(self) if m[0].startswith("read_")
+        )
+        raise ValueError(
+            f"Cannot infer appropriate read function for input, "
+            f"please call one of {msg} directly"
+        )
+
+    def read_csv(
+        self, path: str | Path, table_name: str | None = None, **kwargs: Any
+    ) -> ir.Table:
+        """Register a CSV file as a table in the current database.
+
+        Parameters
+        ----------
+        path
+            The data source. A string or Path to the CSV file.
+        table_name
+            An optional name to use for the created table. This defaults to
+            a sequentially generated name.
+        **kwargs
+            Additional keyword arguments passed to Datafusion loading function.
+
+        Returns
+        -------
+        ir.Table
+            The just-registered table
+        """
+        path = normalize_filename(path)
+        table_name = table_name or f"ibis_read_csv_{next(csv_n)}"
+        # Our other backends support overwriting views / tables when reregistering
+        self._context.deregister_table(table_name)
```

Comment:
I feel like this would be _way_ too aggressive for a non-file-based backend but I think this is ok for Datafusion?

Reply:
@kszucs Thoughts on this?

Reply:
I think this is fine.